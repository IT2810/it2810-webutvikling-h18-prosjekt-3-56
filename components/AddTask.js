import React, { Component } from 'react';
import { Text, View, StyleSheet, TextInput, TouchableOpacity} from 'react-native';


//Component that lets you type in a task and add it to the tasks.
//This is done by doing a callback to TodoScreens addTask function.
export default class AddTask extends Component {
  constructor(props){
    super(props);
    this.state = {
      text: ""
    }
  }
  render() {
    return (

        <View style = {styles.container}>
          <TextInput style = {styles.input} placeHolder = "The road to success is action." onChangeText = {(text) => this.setState({text})}/>
          <TouchableOpacity style = {styles.topac} onPress ={() => {this.state.text !== "" ? this.props.click(this.state.text) : null}}>
            <Text style = {styles.button}>ADD</Text>
          </TouchableOpacity>
        </View>

    );
  }
}

const styles = StyleSheet.create({
  container:{
    display:'flex',
    flexDirection:'row',
    backgroundColor:'#f1c40f'
  },
  input:{
    height:50,
    flex:0.8,
    fontSize:20
  },
  topac:{
    flex:0.2,

  },
  button:{
    height:50,
    backgroundColor:'#f39c12',
    fontSize: 17,
    textAlign:'center',
    paddingTop:14,
    paddingBottom:14
  }
})
