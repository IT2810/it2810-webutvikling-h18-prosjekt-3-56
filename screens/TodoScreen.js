import React from 'react';
import {
  Image,
  Platform,
  ScrollView,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
  AsyncStorage} from 'react-native';
import { ExpoLinksView } from '@expo/samples';
import Task from '../components/Task.js'
import Swipeout from 'react-native-swipeout';

import AddTask from '../components/AddTask.js'



export default class TodoScreen extends React.Component {

  //
  constructor(props){
    super(props);
    this.state= {
      taskList: []
    }
    //has the initial data read been executed?
    this.dataRead = false;

    this.addTaskClicked = this.addTaskClicked.bind(this);
    this.doneClicked = this.doneClicked.bind(this)
  }

  static navigationOptions = {
    title: 'ToDude',
  };

  render() {
    return (
      <View style = {styles.container}>
        <AddTask style = {styles.taskAdder} click = {this.addTaskClicked}/>
        <ScrollView style={styles.scroller}>
          {
            this.state.taskList.map((task) => {
              return(
                <Task key = {task.id} text = {task.text} id = {task.id} click = {this.doneClicked}/>
              )
            })
          }
        </ScrollView>
      </View>
    );
  }

  componentDidMount(){
    console.log("2");
    if(this.dataRead){

    }
    else{
      this.initData();
      console.log("as");
    }
  }
  //Finds the element in the ltruetrueist and removes it from the taskList in state
  doneClicked(id){
    this.setState((prev) => {
      for(let i = 0; i < prev.taskList.length; i++){
        console.log(prev.taskList[i]);
        if (prev.taskList[i].id == id){
          prev.taskList.splice(i,1);
          array = prev.taskList;
          return {taskList: array};
        }
        else{
          console.log("If this prints, you have problem");
        }
      }

    }
    );
  }

  //Adds new task. The id will become id of last element +1 unless the lists length is 0
  addTaskClicked(tex){
    this.setState((prevState) =>
      {
        if (prevState.taskList.length!= 0){
          id = prevState.taskList[prevState.taskList.length -1].id + 1;
          console.log(id);
          return {taskList: [...prevState.taskList, {id: id, text: tex}]}
        }
        else{
          return {taskList: [...prevState.taskList, {id: 0, text: tex}]}
        }
    }, () => {
      this.storeData(this.state);

    }
  )}

  //Reads all data in db with key "tasks" and sets the variable dataRead = true
  initData = async() =>{
    try {
      const value = await AsyncStorage.getItem('tasks');
      if (value !== null) {
        // We have data!!
        tasks = JSON.parse(value);
        console.log(tasks.taskList);
        this.setState(tasks);
        return tasks.taskList;
      }
      else if (value === null) {
        console.log("There was no data stored");
        return {tasks:[]};
      }
      this.dataRead = true;
    } catch (error) {
      console.log(error);
      // Error retrieving data
      return error;
    }
  }

  storeData = async(data) =>{
    try {
      await AsyncStorage.setItem('tasks', JSON.stringify(data));
    } catch (error) {
    // Error saving data
      console.log(error);
    }
  }
}

const styles = StyleSheet.create({
  container:{
    flex:1,
    backgroundColor: '#ecf0f1',
  },
  taskAdder:{

  },
  scroller: {
  }
});
