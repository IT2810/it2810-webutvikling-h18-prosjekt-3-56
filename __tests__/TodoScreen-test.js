import React from 'react';
import AddTask from '../components/AddTask.js';
import TodoScreen from '../screens/TodoScreen.js';
import ScrollView from 'react-native';

//Ettersom TodoScreen er avhengig av komponentene i den, kan vi ikkje
//bare ta en snapshot. Ved hjelp av shallow rendering kan vi teste
//komponenter som er avhengig av subkomponenter.
//tester bare 1 level

import ShallowRenderer from 'react-test-renderer/shallow'
import renderer from 'react-test-renderer';

let rend;
let result;

beforeAll(() => {
  rend = new ShallowRenderer();
  rend.render(<TodoScreen/>);
  result = rend.getRenderOutput();
  console.log(result.props.children[0]);
});

test("Shallow render test of Todoscreen",() =>{
  expect(result.ref).toBe(null);
});
