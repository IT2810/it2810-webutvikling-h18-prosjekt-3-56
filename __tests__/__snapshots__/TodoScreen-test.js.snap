// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ecf0f1",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#f1c40f",
        "display": "flex",
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeHolder="The road to success is action."
      style={
        Object {
          "flex": 0.8,
          "fontSize": 20,
          "height": 50,
        }
      }
    />
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 0.2,
          "opacity": 1,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "backgroundColor": "#f39c12",
            "fontSize": 17,
            "height": 50,
            "paddingBottom": 14,
            "paddingTop": 14,
            "textAlign": "center",
          }
        }
      >
        ADD
      </Text>
    </View>
  </View>
</View>
`;
